### PR TITLE
Add retry logic to cover deadlock situations during move operations

### PR DESCRIPTION
**Pull request by artonge**
Should help with https://github.com/nextcloud/server/issues/43759

### CODE REVIEW

File: lib/private/Files/Cache/Cache.php [hunk 1]
```diff
@@ -715,12 +715,29 @@ public function moveFromCache(ICache $sourceCache, $sourcePath, $targetPath) {
...
+						$this->connection->rollBack();
+						throw $e;
+					} catch (RetryableException $e) {
+						// Simply throw if we already retried 4 times.
+						if ($i === 4) {
+							throw $e;
+						}
+
+						// Sleep for 1 seconde to give some time to the other transaction to finish.
+						sleep(1);
```
Comment (icewind1991) [inline lib/private/Files/Cache/Cache.php:735]: I expect that there isn't a need to wait here, the deadlock error should clear both waiting locks I believe.

Comment (icewind1991) [reply lib/private/Files/Cache/Cache.php:735]: Or I would at least reduce the sleep to 10-100ms

Comment (Altahrim) [reply lib/private/Files/Cache/Cache.php:736]: I think it doesn't clear the original lock but a much shorter sleep should be enough.
We can also have an optimistic retry and wait longer after:
```suggestion
						usleep($i*$i*100);
```
I also think we should rollback before waiting

File: lib/private/Files/Cache/Cache.php [hunk 2]
```diff
@@ -715,12 +715,31 @@ public function moveFromCache(ICache $sourceCache, $sourcePath, $targetPath) {
...
+					} catch (RetryableException $e) {
+						// Simply throw if we already retried 4 times.
+						if ($i === $retryLimit) {
+							throw $e;
+						}
+
+						$this->connection->rollBack();
+
+						// Sleep a bit to give some time to the other transaction to finish.
+						sleep(100 * $i);
```
Comment (icewind1991) [inline lib/private/Files/Cache/Cache.php:738]: I think waiting for up to 400s isn't what you mean here :see_no_evil:

Comment (artonge) [reply lib/private/Files/Cache/Cache.php]: AH ! thanks :)

File: lib/private/Files/Cache/Cache.php [hunk 3]
```diff
@@ -715,12 +715,31 @@ public function moveFromCache(ICache $sourceCache, $sourcePath, $targetPath) {
...
+					} catch (RetryableException $e) {
+						// Simply throw if we already retried 4 times.
+						if ($i === $retryLimit) {
+							throw $e;
+						}
+
+						$this->connection->rollBack();
+
+						// Sleep a bit to give some time to the other transaction to finish.
+						usleep(100 * $i);
```
Comment (icewind1991) [inline lib/private/Files/Cache/Cache.php:738]: Now you're only sleeping for up to 0.4ms :)

`usleep(100 * 1000 * $i);` is what you're looking for

File: lib/private/Files/Cache/Cache.php [hunk 4]
```diff
@@ -715,12 +715,31 @@ public function moveFromCache(ICache $sourceCache, $sourcePath, $targetPath) {
...
-				} catch (\OC\DatabaseException $e) {
-					$this->connection->rollBack();
-					throw $e;
+				// Retry transaction in case of RetryableException like deadlocks.
+				// Retry up to 4 times because we should receive up to 4 concurrent requests from the frontend
+				$retryLimit = 4;
+				for ($i = 1; $i <= $retryLimit; $i++) {
+					try {
+						$this->connection->beginTransaction();
+						$query->execute();
```
Comment (come-nc) [inline lib/private/Files/Cache/Cache.php:724]:
```suggestion
						$query->executeStatement();
```

**Review by come-nc — COMMENTED** [review]
I think I like it, especially if it works? Did not test yet.

Is there anyway to let the DB or our Query layer handle that? Have some kind of generic way of saying «run this query in a transaction, if it deadlocks retry up to X times» ?

File: lib/private/Files/Cache/Cache.php (hunk 4 above)
Comment (artonge) [reply lib/private/Files/Cache/Cache.php:724]: Indeed, let see if it works as we expect it to, and then we can maybe add an argument to `executeStatement()`

Comment (artonge) [reply lib/private/Files/Cache/Cache.php:724]: I also apply your suggestion on the `execute` call

**Review by Altahrim — APPROVED** [review]

File: lib/private/Files/Cache/Cache.php (hunk 3 above)
Comment (artonge) [reply lib/private/Files/Cache/Cache.php]: Well, thanks for the handholding :see_no_evil:

File: lib/private/Files/Cache/Cache.php [hunk 5]
```diff
@@ -715,12 +715,31 @@ public function moveFromCache(ICache $sourceCache, $sourcePath, $targetPath) {
...
 				}
 
-				try {
-					$query->execute();
-				} catch (\OC\DatabaseException $e) {
-					$this->connection->rollBack();
-					throw $e;
+				// Retry transaction in case of RetryableException like deadlocks.
+				// Retry up to 4 times because we should receive up to 4 concurrent requests from the frontend
+				$retryLimit = 4;
```
Comment (juliushaertl) [inline lib/private/Files/Cache/Cache.php:720]: https://github.com/nextcloud/server/blob/050c6d53b38d6737a49c13d4e6a20a6062160503/lib/public/AppFramework/Db/TTransactional.php#L94 could be used for that though it does not have a wait in it

**Review by juliusknorr — APPROVED** [review]

File: lib/private/Files/Cache/Cache.php (hunk 5 above)
Comment (artonge) [reply lib/private/Files/Cache/Cache.php:720]: Oh, seems better to use that, no? We could even add a support for an optional wait time.

Comment (artonge) [reply lib/private/Files/Cache/Cache.php:720]: But is a trait, I would rather have that in IDBConnection or IQueryBuilder

**Pull request merged**